### ISSUE TITLE
Segmentation fault: 11 during pandas to_csv operation

### ISSUE THREAD

**Issue by berthelemy** [pydata/pandas#11174]
Following a suggestion on [StackOverflow](http://stackoverflow.com/questions/32720514/segmentation-fault-11-during-pandas-to-csv-operation), I'm posting this here:

I'm working on Mac OS 10.10.5, using a Python 2.7 virtualenv and pandas installed via pip.

I have a Pandas dataFrame (called 'pivot') which is 13843 rows by 40 columns. Most of the data cells are single characters (A, B, C or D).

I am trying to export this to a csv file using the command:

    pivot.to_csv(path_or_buf="test.csv",sep=",", encoding='utf-8')

The output CSV gets to row 8743, which is truncated after the 8th column, and the program crashes out with a Segmentation fault: 11

It shouldn't be a memory problem, as my Mac has 16Gb RAM.

Result of pandas.show_versions() : 

INSTALLED VERSIONS
------------------
commit: None
python: 2.7.10.final.0
python-bits: 64
OS: Darwin
OS-release: 14.5.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: None
LANG: en_GB.UTF-8

pandas: 0.16.2
nose: None
Cython: None
numpy: 1.9.2
scipy: None
statsmodels: None
IPython: None
sphinx: None
patsy: None
dateutil: 2.4.2
pytz: 2015.4
bottleneck: None
tables: None
numexpr: None
matplotlib: None
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: None
httplib2: None
apiclient: None
sqlalchemy: None
pymysql: None
psycopg2: None

Here's a sample of the dataFrame I'm trying to export. Note that the first two columns are indexes.

CandidateRegistrationID | ExamDate | 301|  302|  ...|  331 | 332 |  333 | 334 | 340 |
----------------------------------|---------------|------|------|-----|------|--------|-------|-------|-------|
abc123 |              2015-01-18 |   B |    A | ...  |   C |   C |    B |    A |    B |   
abc124  |           2015-02-16 |   C |    A | ... |    C |    C |    B |    A |    B |   

The minimum reproducing sample is a dataFrame of size: pivot.iloc[0:8774, 00:40]

Anything with fewer rows than that exports OK. The number of columns makes no difference. The segmentation fault appears even with one column.

**Comment by dsm054** [comment]
The same suggestion made on SO applies here. :-)   Could you try to find the minimum reproducing case? 

(Also, if you could edit your description to include the result of `pd.show_versions()`, that would be helpful as well.)

**Comment by berthelemy** [comment]
Thanks... I was working my way up to that - trying to isolate the issue... :-)

**Comment by berthelemy** [comment]
Updated main description with an example of the dataFrame and details of minimum reproducing size.

**Comment by jreback** [comment]
show the minimum that fails
and df.info()

if u can pickle the frame and post a link would be great

**Comment by berthelemy** [comment]
Here's the df.info() output:

<class 'pandas.core.frame.DataFrame'>
Int64Index: 553760 entries, 0 to 553759
Data columns (total 4 columns):
CandidateRegistrationID    553760 non-null object
ExamDate                   553760 non-null object
QuestionCode               553760 non-null object
CandidateResponse          553760 non-null object
dtypes: object(4)
memory usage: 21.1+ MB

I'm afraid I can't pickle and publish the frame, as it contains confidential CandidateRegistrationID data, and my pandas/python skills aren't up to obfuscating one index column in a multiIndex dataFrame.

**Comment by jorisvandenbossche** [comment]
@berthelemy you say the minimum reproducible size is `pivot.iloc[0:8774, 00:40]`. But is it the number of rows that is the problem, or those particular rows? Eg what happens in you use `pivot.iloc[1000:9774, 00:40]`? Or is is eg specifically that last row of the subselection that causes the problem?

**Comment by berthelemy** [comment]
This is interesting...

With:

    pivot = pivot.iloc[8770:8773, 00:1]

It gives:

QuestionCode                                   301 
CandidateRegistrationID ExamDate       
abc123              2013-02-25               C
abc124              2013-02-25               C
abc125              2013-02-25               C

With:

    pivot = pivot.iloc[8770:8774, 00:1]

Just printing the dataFrame gives:

Segmentation fault: 11

With:

    pivot = pivot.iloc[8774:8777, 00:1]

It gives:

QuestionCode                               301 
CandidateRegistrationID ExamDate       
abc125              2013-02-25    C
abc126              2013-02-25    C
abc127              2013-02-25    C

So, it seems like there might be a problem with the data at this point. But I'm not sure how to check it without being able to print?

**Comment by jorisvandenbossche** [comment]
What if you access the numpy array, does it then segfault?

```
pivot.iloc[8773, 00:1].values
```

And does it segfault if you *print* it (with `print`) or also if you just display it?

**Comment by berthelemy** [comment]
On it's own:

    pivot.iloc[8773, 00:1].values

Gives nothing.

    print (pivot.iloc[8773, 00:1].values)

Gives segmentation fault: 11

**Comment by jorisvandenbossche** [comment]
What do you mean with 'nothing'? Eg in IPython there is not Out[] prompt?

**Comment by berthelemy** [comment]
I'm using a script with output to the console. By nothing, I mean a blank space.

    df=pandas.DataFrame(data, columns=['CandidateRegistrationID','ExamDate','QuestionCode','CandidateResponse'])

    pivot = df.pivot_table(index=['CandidateRegistrationID','ExamDate'],columns='QuestionCode',values='CandidateResponse',aggfunc = lambda x: x)

    pivot = pivot.iloc[8770:8773, 00:1]
    pivot # [Out] = blank

    #print (pivot) # [Out] = Segmentation fault: 11

**Comment by jorisvandenbossche** [comment]
Yes of course, from a script you will see no output when you do not print it. But can you try it in a interactive console to see what it outputs? Just to be able to inspect the objects on that row (there no printing is needed)

**Comment by TomAugspurger** [comment]
Also maybe try it from a Notebook incase it's the terminal that can't print a unicode character or something like that.

**Comment by berthelemy** [comment]
# Using a Notebook:

In[1]:

    from __future__ import print_function

    import sys
    import pandas
    import numpy
 
    filepath = sys.argv[1]

    filepath = 'data/examdata.csv'
    print("Importing data...")
    data_df = pandas.read_csv(filepath,low_memory=False)

    #print (data_df.columns)
    print("Filtering based on UniquePaperNo...")
    filtered_df = data_df.loc[data_df['UniquePaperNo'] == 211]
    
    #print(filtered_df)
    print("Pivoting...")
    pivot = filtered_df.pivot_table(index=['CandidateRegistrationID','ExamDate'],columns='QuestionCode',values='CandidateResponse',aggfunc = lambda x: x)

    #print(pivot)
    pivot2 = pivot.iloc[8773, 00:1].values
    #print("Exporting...")
    #pivot.to_csv(path_or_buf="test.csv",sep=",", encoding='utf-8') # export data
    print("Done.")

Out[1]:
Importing data...
Filtering based on UniquePaperNo...
Pivoting...
Done.

In[2]:
pivot2

Out[2]:
The kernel appears to have died. It will restart automatically.

# Using the python interpreter:

In[1] - as above
In[2] - as above
Out[2] - Segmentation fault: 11

**Comment by berthelemy** [comment]
I have tried to pickle the dataFrame, but that also failed with a segfault...

I also ran a trace (using dtruss), but the output just stopped during the csv export with no indication of what was going on. Here's the last line (CandidateRegistrationID has been changed):

write_nocancel(0x3, "A,A,B,C,B,A,A,C,D,D,A,B,B,A,B,D,B,C,C,B,A,B,C,C,C,D,A\abcd1234,2013-02-13,C,A,B,A,C,B,A,D,B,A,C,A,D,C,A,B,C,B,A,A,C,D,D,C,B,B,A,B,D,A,C,C,B,A,B,C,C,B,D,A\abcd1235,2013-02-13,C,A,B,A,C,B,A,D,B,A,C,A,D,C,A,B,C,B,A,A,C,D,D,C,B,B,A,B,D,A,C,C,B,A,B,C,C,B,D,", 0x1000)		 = 4096 0

**Comment by kawochen** [comment]
@berthelemy  Can you also show some obfuscated version of the corresponding line in the csv? On what platform was the csv created?

**Comment by berthelemy** [comment]
@kawochen The CSV was created in a bespoke .aspx application.

The raw CSV imports into a dataFrame OK, and exports directly out to csv OK too.

It's only after the pivot_table operation that the problems occur.

I have tracked the data through to identify the line in the dataFrame that only partially exports.

It is row number 8743, which comprises:
```
testUserID,2013-02-23,C,A,D,A,C,B,A,D,B,A,B,A,D,A,A,A,D,B,A,A,C,A,D,C,B,B,
```

As you can see, this truncates after 26 grade letters, whereas the raw data (pre-pivoting) for this userID contains 40 lines, each with a grade letter

```
Id,CandidateRegistrationID,CandidateID,OurReference,QualificationCode,ExamCode,ExamDate,QualificationName,DataSource,QuestionNo,CandidateResponse,CorrectAnswerChoice,UniquePaperNo,QuestionCode
1372783,testUserID,testUserIDno,50421,SWT     ,SWT ,2012-01-01,Exam name XYZ,PR  ,1,C,C,211,301 
1372784,testUserID,testUserIDno,50421,SWT     ,SWT ,2012-01-01,Exam name XYZ,PR  ,2,A,A,211,302 
1372785,testUserID,testUserIDno,50421,SWT     ,SWT ,2012-01-01,Exam name XYZ,PR  ,3,D,B,211,303 
1372786,testUserID,testUserIDno,50421,SWT     ,SWT ,2012-01-01,Exam name XYZ,PR  ,4,A,A,211,304 
1372787,testUserID,testUserIDno,50421,SWT     ,SWT ,2012-01-01,Exam name XYZ,PR  ,5,C,C,211,305 
1372788,testUserID,testUserIDno,50421,SWT     ,SWT ,2012-01-01,Exam name XYZ,PR  ,6,B,B,211,306 
1372789,testUserID,testUserIDno,50421,SWT     ,SWT ,2012-01-01,Exam name XYZ,PR  ,7,A,A,211,307 
1372790,testUserID,testUserIDno,50421,SWT     ,SWT ,2012-01-01,Exam name XYZ,PR  ,8,D,D,211,308 
1372791,testUserID,testUserIDno,50421,SWT     ,SWT ,2012-01-01,Exam name XYZ,PR  ,9,B,B,211,309 
1372792,testUserID,testUserIDno,50421,SWT     ,SWT ,2012-01-01,Exam name XYZ,PR  ,10,A,A,211,310 
1372793,testUserID,testUserIDno,50421,SWT     ,SWT ,2012-01-01,Exam name XYZ,PR  ,11,B,C,211,311 
1372794,testUserID,testUserIDno,50421,SWT     ,SWT ,2012-01-01,Exam name XYZ,PR  ,12,A,A,211,312 
1372795,testUserID,testUserIDno,50421,SWT     ,SWT ,2012-01-01,Exam name XYZ,PR  ,13,D,D,211,313 
1372796,testUserID,testUserIDno,50421,SWT     ,SWT ,2012-01-01,Exam name XYZ,PR  ,14,A,C,211,314 
1372797,testUserID,testUserIDno,50421,SWT     ,SWT ,2012-01-01,Exam name XYZ,PR  ,15,A,A,211,315 
1372798,testUserID,testUserIDno,50421,SWT     ,SWT ,2012-01-01,Exam name XYZ,PR  ,16,A,B,211,316 
1372799,testUserID,testUserIDno,50421,SWT     ,SWT ,2012-01-01,Exam name XYZ,PR  ,17,D,C,211,317 
1372800,testUserID,testUserIDno,50421,SWT     ,SWT ,2012-01-01,Exam name XYZ,PR  ,18,B,B,211,318 
1372801,testUserID,testUserIDno,50421,SWT     ,SWT ,2012-01-01,Exam name XYZ,PR  ,19,A,A,211,319 
1372802,testUserID,testUserIDno,50421,SWT     ,SWT ,2012-01-01,Exam name XYZ,PR  ,20,A,A,211,320 
1372803,testUserID,testUserIDno,50421,SWT     ,SWT ,2012-01-01,Exam name XYZ,PR  ,21,C,C,211,321 
1372804,testUserID,testUserIDno,50421,SWT     ,SWT ,2012-01-01,Exam name XYZ,PR  ,22,A,D,211,322 
1372805,testUserID,testUserIDno,50421,SWT     ,SWT ,2012-01-01,Exam name XYZ,PR  ,23,D,D,211,323 
1372806,testUserID,testUserIDno,50421,SWT     ,SWT ,2012-01-01,Exam name XYZ,PR  ,24,C,C,211,324 
1372807,testUserID,testUserIDno,50421,SWT     ,SWT ,2012-01-01,Exam name XYZ,PR  ,25,B,B,211,325 
1372808,testUserID,testUserIDno,50421,SWT     ,SWT ,2012-01-01,Exam name XYZ,PR  ,26,B,B,211,326 
1372809,testUserID,testUserIDno,50421,SWT     ,SWT ,2012-01-01,Exam name XYZ,PR  ,27,A,A,211,327 
1372810,testUserID,testUserIDno,50421,SWT     ,SWT ,2012-01-01,Exam name XYZ,PR  ,28,B,B,211,328 
1372811,testUserID,testUserIDno,50421,SWT     ,SWT ,2012-01-01,Exam name XYZ,PR  ,29,D,D,211,329 
1372812,testUserID,testUserIDno,50421,SWT     ,SWT ,2012-01-01,Exam name XYZ,PR  ,30,A,A,211,330 
1372813,testUserID,testUserIDno,50421,SWT     ,SWT ,2012-01-01,Exam name XYZ,PR  ,31,C,C,211,331 
1372814,testUserID,testUserIDno,50421,SWT     ,SWT ,2012-01-01,Exam name XYZ,PR  ,32,C,C,211,332 
1372815,testUserID,testUserIDno,50421,SWT     ,SWT ,2012-01-01,Exam name XYZ,PR  ,33,B,B,211,333 
1372816,testUserID,testUserIDno,50421,SWT     ,SWT ,2012-01-01,Exam name XYZ,PR  ,34,A,A,211,334 
1372817,testUserID,testUserIDno,50421,SWT     ,SWT ,2012-01-01,Exam name XYZ,PR  ,35,D,B,211,335 
1372818,testUserID,testUserIDno,50421,SWT     ,SWT ,2012-01-01,Exam name XYZ,PR  ,36,C,C,211,336 
1372819,testUserID,testUserIDno,50421,SWT     ,SWT ,2012-01-01,Exam name XYZ,PR  ,37,A,C,211,337 
1372820,testUserID,testUserIDno,50421,SWT     ,SWT ,2012-01-01,Exam name XYZ,PR  ,38,D,B,211,338 
1372821,testUserID,testUserIDno,50421,SWT     ,SWT ,2012-01-01,Exam name XYZ,PR  ,39,D,D,211,339 
1372822,testUserID,testUserIDno,50421,SWT     ,SWT ,2012-01-01,Exam name XYZ,PR  ,40,A,A,211,340
```

I can't see anything in this part of the data file that might be causing the problem.

I have tried creating a pivoted dataFrame from another segment of the raw data, which only has 1000 or so rows, and that gives the same segfault, with a truncated csv file - but this time truncated at row 80.

**Comment by kawochen** [comment]
It sounds like it's a problem with 8774 or the end of 8773.  Looking at the csv output is sometimes inaccurate since it could be a buffered writer, though I've not looked at the code to make sure.

**Comment by berthelemy** [comment]
@kawochen I've discovered this is happening on other slices of the data. Small slices work fine, but large slices fail. I'm not able to isolate it down to any particular lines of data, it seems to be related to the size. Although, one slice failed after 80 lines, where another slice failed after 8000...

I'm looking at the raw csv inside Sublime Text 2, and seeing no difference in the lines where it fails from lines that work.

I'm going to leave this issue open in case anyone has any more ideas, but I've run out of time to do any more work on it, and have accepted that, in this case the pivot_table function won't work for me.

**Comment by jreback** [comment]
@berthelemy well, w/o posting the data this is not possible to solve. If you are able to post a reproducible example then pls reopen.

**Comment by berthelemy** [comment]
Hi all,

I think I've isolated the problem. It's an issue with the data.

I was working on the assumption that the same exam could not be taken more than once on the same day. This turns out not to be true.

I found this out by isolating one particular candidate's data which was crashing and trying to pivot that.

The error given was ValueError('Function does not reduce')

Thanks for all your help. I'll go back to the data.

Mark
